### ISSUE TITLE
[bug] Config to use the firstuseauthentificator in qhub

### ISSUE THREAD

**Issue by MarineChap** [Quansight/qhub#831]
(follow-up issue after asking this question of gitter)

I am trying to use the firstuseauthentificator in qhub.
my configuration looks like this : 
```
security:
  authentication:
    type: custom
    authentication_class: 'firstuseauthenticator.FirstUseAuthenticator'
    config: 
       min_password_length: 5

```
Then I run `python -m qhub deploy --config qhub-config.yaml` to update the deployment. (by the way is that the command I should use to update the deployment ? the doc is not super clear for this step.)

And the error displayed this : 
```
[terraform]: module.qhub.module.kubernetes-jupyterhub.helm_release.jupyterhub: Still modifying... [id=jupyterhub, 5m0s elapsed]
[terraform]: module.qhub.module.kubernetes-jupyterhub.helm_release.jupyterhub: Still modifying... [id=jupyterhub, 5m10s elapsed]
[terraform]: module.qhub.module.kubernetes-jupyterhub.helm_release.jupyterhub: Still modifying... [id=jupyterhub, 5m20s elapsed]
[terraform]: module.qhub.module.kubernetes-jupyterhub.helm_release.jupyterhub: Still modifying... [id=jupyterhub, 5m30s elapsed]
[terraform]: module.qhub.module.kubernetes-jupyterhub.helm_release.jupyterhub: Still modifying... [id=jupyterhub, 5m40s elapsed]
[terraform]: 
[terraform]: Error: timed out waiting for the condition
[terraform]: 
[terraform]:   on modules/kubernetes/services/jupyterhub/main.tf line 6, in resource "helm_release" "jupyterhub":
[terraform]:    6: resource "helm_release" "jupyterhub" {
[terraform]: 
[terraform]: 
```

Some advice on the config I should use would be nice ? Thanks !

Additional note if it can help : 
This is my config with the jupyterhub helm repo and it works: 
```
  hub:
    config:
     JupyterHub:
          authenticator_class: firstuseauthenticator.FirstUseAuthenticator
     FirstUseAuthenticator:
          min_password_length: 5
          admin_access: true
          admin_users: 
          - marine
```
## Your environment
qhub  == 0.3.12
minkube with kubernetes-version=1.19.4
jupyterhub-firstuseauthenticator == 0.14.1 (https://github.com/jupyterhub/firstuseauthenticator)

Thanks

**Comment by viniciusdc** [comment]
Hi @MarineChap, thanks for utilizing QHub and opening this issue. I've made some attempts to reproduce the above situation, here's what I found from jupyterhub service:
```
[C 2021-09-30 15:14:43.639 JupyterHub application:89] Bad config encountered during initialization: 
The 'authenticator_class' trait of <jupyterhub.app.JupyterHub object at 0x7f62e3926950> instance 
must be a type, but 'firstuseauthenticator.FirstUseAuthenticator' could not be imported
```
@costrouc Any thoughts on this matter?

**Comment by MarineChap** [comment]
Hi @viniciusdc  
Sorry for not being clear, I had this error too but this is fixed by installing the the package in your environment.
`pip install jupyterhub-firstuseauthenticator`

**Comment by iameskild** [comment]
Hey @viniciusdc and @danlester, do either of you know if this issue will be resolved by the recent addition of keycloak?

**Comment by danlester** [comment]
@iameskild Actually, this will no longer be supported directly:
```
security:
  authentication:
    type: custom
```
which is something we need to add to the release notes. I think it would be a good idea to create a draft release so we can add information like that when we come to it, or would you prefer changelog to be in the docs directly?

The `qhub upgrade` command should also give some kind of warning if it comes across this configuration,
I made an issue for that here: #931 

I would be happy to solve that issue, although it could be useful experience for someone to see the upgrade process.

In terms of how to achieve the desired authentication behavior (i.e. something similar to First Use Authenticator), Keycloak would need to be configured for this if possible.

**Comment by iameskild** [comment]
Thanks a lot @danlester for the update! I have created a [draft release notes page](https://github.com/Quansight/qhub/releases/tag/untagged-c768b51c28e44dbcbbba) for us; I added a note about this issue but perhaps it needs more detail. 

I agree that the `qhub upgrade` command is going to be quite important. At the moment, it's not slated to be included in the 0.4.0 release but perhaps we can advocate that it be added to an upcoming release.

**Comment by danlester** [comment]
To be clear, `qhub upgrade` is already in `main` as part of Keycloak (and is essential for it in my opinion).

**Comment by danlester** [comment]
Just coming back to the use of FirstUseAuthenticator by @MarineChap: as explained JupyterHub authenticators won't be a part of the upcoming v0.4 release of QHub, so the config including FirstUseAuthenticator will not be supported directly.

It might depend on exactly how you have been configuring FirstUseAuthenticator, but a similar flow using the new Keycloak system would be to ask users to register. They can choose a username and password at that time.

To be clear, this is only available in the `main` (development) branch of QHub at the moment, but should be released as a full version soon (v0.4). This is how you would enable registration:

![Screenshot 2021-12-03 at 13 51 45](https://user-images.githubusercontent.com/786238/144615014-261beb79-a760-4c99-978d-ac957691d96f.png)

To allow users to reset their passwords by email you would also need to add details for an SMTP Email server in the Email tab in that screenshot.

**Comment by MarineChap** [comment]
Ok sound good for me. I will wait for the release and test it. 
Thank you for your work. 

Have a good day
